File: LeetCode/LeetCode/Problems/5/Problem515.cs
using LeetCode.DataStructures;
using System.Collections.Generic;
using System;
using System.Linq;

public class Problem515
{
    public IList<int> LargestValues(TreeNode root)
    {
        var q = new Queue<TreeNode>();
        if (root is not null)
            q.Enqueue(root);

        var res = new List<int>();

        while (q.Any())
        {
            var count = q.Count();

            var max = int.MinValue;
            while (count > 0)
            {
                count--;
                var current = q.Dequeue();
                max = Math.Max(max, current.val);
                if (current.left is not null)
                    q.Enqueue(current.left);
                if (current.right is not null)
                    q.Enqueue(current.right);
            }

            res.Add(max);
        }

        return res;
    }
}

File: leetcode_rust/src/problems/problem_515.rs
use crate::structures::tree_node::TreeNode;
use std::cell::RefCell;
use std::collections::VecDeque;
use std::rc::Rc;

pub fn largest_values(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
    let mut res = Vec::new();

    if let Some(node) = root {
        let mut q = VecDeque::new();
        q.push_back(node);

        while !q.is_empty() {
            let mut count = q.len();
            let mut max = i32::MIN;
            while count > 0 {
                count -= 1;
                if let Some(node) = q.pop_front() {
                    let current = node.borrow();
                    max = max.max(current.val);
                    if let Some(node) = current.left.clone() {
                        q.push_back(node);
                    }
                    if let Some(node) = current.right.clone() {
                        q.push_back(node);
                    }
                }
            }

            res.push(max);
        }
    }

    return res;
}
